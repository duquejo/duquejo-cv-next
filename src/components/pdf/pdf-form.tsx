import { Button } from '@/components/ui/button';
import { Separator } from '@/components/ui/separator';
import { FormEvent, useState } from 'react';

interface Props {
  onSubmitFinish: () => void;
  button: string;
  buttonLoading: string;
}

export const PdfForm = ({ onSubmitFinish, button, buttonLoading }: Props) => {
  const [isLoading, setIsLoading] = useState(false);

  const handleSubmit = async (event: FormEvent) => {
    event.preventDefault();
    setIsLoading(true);

    try {
      const response = await fetch('/api/v1/pdf', {
        method: 'POST',
        next: { revalidate: 3600 },
      });

      if (response.status !== 200) {
        throw new Error('Request failed with status code ' + response.status);
      }

      const blob = await response.blob();
      const url = window.URL.createObjectURL(blob);
      const link = document.createElement('a');
      link.href = url;
      link.download = 'cv_jose_duque.pdf';
      link.click();
      link.remove();

      window.URL.revokeObjectURL(url);
      onSubmitFinish();
    } catch (e: unknown) {
      if (e instanceof Error) {
        console.error(e.message);
        return;
      }
    } finally {
      setIsLoading(false);
    }
  };

  return (
    <form onSubmit={handleSubmit} className="flex flex-col">
      <Separator orientation="horizontal" className="mt-2 mb-3" />
      <Button type="submit" size="sm" disabled={isLoading}>
        {isLoading ? buttonLoading : button}
      </Button>
    </form>
  );
};
