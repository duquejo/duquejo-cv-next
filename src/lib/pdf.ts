import type { Template } from '@pdfme/common';
import { BASE_PDF } from '@/lib';

export const getBaseTemplate: Template = {
  basePdf: BASE_PDF,
  schemas: [
    [
      {
        name: 'introduction',
        type: 'text',
        position: {
          x: 20.05,
          y: 77.06,
        },
        width: 169.9,
        height: 44.99,
        alignment: 'center',
        fontSize: 13,
        lineHeight: 1.5,
      },
      {
        name: 'experience',
        type: 'text',
        position: {
          x: 46.64,
          y: 150.14,
        },
        width: 55,
        height: 8,
        fontSize: 15,
        lineHeight: 1,
      },
      {
        name: 'languages',
        type: 'text',
        position: {
          x: 138.54,
          y: 144.02,
        },
        lineHeight: 1.3,
        width: 70,
        height: 13.03,
        fontSize: 12,
      },
      {
        name: 'databases',
        type: 'text',
        position: {
          x: 20,
          y: 191,
        },
        width: 80,
        height: 13,
        fontSize: 10,
        lineHeight: 1.3,
      },
      {
        name: 'cloud',
        type: 'text',
        position: {
          x: 110,
          y: 191,
        },
        width: 80,
        height: 13,
        fontSize: 10,
        lineHeight: 1.3,
      },
      {
        name: 'tools',
        type: 'text',
        position: {
          x: 110,
          y: 191,
        },
        width: 80,
        height: 13,
        fontSize: 10,
        lineHeight: 1.3,
      },
      {
        name: 'backend',
        type: 'text',
        position: {
          x: 20,
          y: 210,
        },
        width: 80,
        height: 13,
        fontSize: 10,
        lineHeight: 1.3,
      },
      {
        name: 'cicd',
        type: 'text',
        position: {
          x: 110,
          y: 210,
        },
        width: 80,
        height: 13,
        fontSize: 10,
        lineHeight: 1.3,
      },
      {
        name: 'frontend',
        type: 'text',
        position: {
          x: 20,
          y: 234,
        },
        width: 80,
        height: 13,
        fontSize: 10,
        lineHeight: 1.3,
      },
      {
        name: 'programming_languages',
        type: 'text',
        position: {
          x: 110,
          y: 234,
        },
        width: 80,
        height: 8,
        fontSize: 10,
        lineHeight: 1.3,
      },
      {
        name: 'architecture',
        type: 'text',
        position: {
          x: 20,
          y: 257,
        },
        width: 80,
        height: 13,
        fontSize: 10,
        lineHeight: 1.3,
      },
      {
        name: 'methodologies',
        type: 'text',
        position: {
          x: 110,
          y: 262,
        },
        width: 80,
        height: 5,
        fontSize: 10,
        lineHeight: 1.3,
      },
      {
        name: 'security',
        type: 'text',
        position: {
          x: 110,
          y: 248,
        },
        width: 80,
        height: 5,
        fontSize: 10,
        lineHeight: 1.3,
      },
      {
        name: 'email_bottom',
        type: 'text',
        position: {
          x: 147,
          y: 284.2,
        },
        width: 41.63,
        height: 5,
        fontSize: 10,
        lineHeight: 1.3,
      },
      {
        name: 'website_bottom',
        type: 'text',
        position: {
          x: 147,
          y: 278.9,
        },
        width: 50,
        height: 5,
        fontSize: 10,
        lineHeight: 1.3,
      },
    ],
  ],
};
