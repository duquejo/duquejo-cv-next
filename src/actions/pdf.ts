'use server';

import type { LanguageType } from '@/interfaces';
import { calculateYears, getBaseTemplate } from '@/lib';

const years = calculateYears();

const PDF_EMAIL = process.env.PDF_EMAIL || '';
const PDF_WEBSITE = process.env.PDF_WEBSITE || '';

export async function generatePdf() {
  const [getTranslations, generate] = await Promise.all([
    import('next-intl/server').then((module) => module.getTranslations),
    import('@pdfme/generator').then((module) => module.generate),
  ]);

  const t = await getTranslations('Pdf');

  return generate({
    inputs: [
      {
        introduction: t('introduction'),
        databases: t('databases'),
        cicd: t('cicd'),
        backend: t('backend'),
        programming_languages: t('programming_languages'),
        frontend: t('frontend'),
        cloud: t('cloud'),
        security: t('security'),
        architecture: t('architecture'),
        methodologies: t('methodologies'),
        experience: t('experience', { years }),
        languages: t
          .raw('languages')
          .map((lang: LanguageType) => `${lang.title} - ${lang.subtitle}`)
          .join('\n'),
        email_bottom: PDF_EMAIL,
        website_bottom: PDF_WEBSITE,
      },
    ],
    template: getBaseTemplate,
  }).catch((error) => {
    console.error(error);
    return Promise.reject(error);
  });
}
