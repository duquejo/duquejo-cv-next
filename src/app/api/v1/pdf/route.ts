import { NextRequest, NextResponse } from 'next/server';
import { generatePdf } from '@/actions/pdf';

const parseAllowedOrigins = () => {
  if (!process.env.ALLOWED_ORIGINS) return [];

  return process.env.ALLOWED_ORIGINS.split(',').map((url) => String(url).trim());
};

const corsOptions = {
  'Access-Control-Allow-Methods': 'POST, OPTIONS',
  'Access-Control-Allow-Headers': 'Origin, X-Requested-With, Content-Type, Accept, Authorization',
};

const pdfOptions = () => {
  const fileName = process.env.PDF_FILENAME || '';
  return {
    'Content-Disposition': `attachment; filename=${fileName}.pdf`,
    'Content-Type': 'application/pdf',
  };
};

export async function POST(request: NextRequest): Promise<Response> {
  const origin = request.headers.get('origin') ?? '';
  const isAllowedOrigin = parseAllowedOrigins().includes(origin);

  const concatOptions: Record<string, string> = {
    ...corsOptions,
    ...pdfOptions(),
  };

  if (isAllowedOrigin) {
    concatOptions['Access-Control-Allow-Origin'] = origin;
  }

  try {
    const pdfBuffer = await generatePdf();

    return new Response(pdfBuffer, {
      status: 200,
      headers: concatOptions,
    });
  } catch (error) {
    console.error(error);
    return new Response('The PDF cannot be generated, try later.', {
      status: 500,
      headers: {
        'Content-Type': 'application/json',
      },
    });
  }
}

export async function OPTIONS(request: NextRequest): Promise<Response> {
  const origin = request.headers.get('origin') ?? '';
  const isAllowedOrigin = parseAllowedOrigins().includes(origin);

  const preflightHeaders = {
    ...(isAllowedOrigin && { 'Access-Control-Allow-Origin': origin }),
    ...corsOptions,
  };
  return NextResponse.json({}, { headers: preflightHeaders });
}
