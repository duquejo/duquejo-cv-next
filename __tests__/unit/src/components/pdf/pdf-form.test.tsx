import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import { PdfForm } from '@/components/pdf/pdf-form';
import { beforeEach, type Mock, vi } from 'vitest';

vi.mock('next/router', () => ({
  useRouter: () => ({
    push: vi.fn(),
  }),
}));

describe('<PdfForm /> tests', () => {
  const mockedCallback = vi.fn();

  const args = {
    button: 'Mocked button',
    buttonLoading: 'Mocked loading',
    onSubmitFinish: mockedCallback,
  };

  beforeEach(() => {
    vi.clearAllMocks();

    global.fetch = vi.fn().mockImplementation(() =>
      Promise.resolve({
        status: 200,
        blob: () => Promise.resolve(new Blob(['testing'], { type: 'application/pdf' })),
      }),
    ) as Mock;

    global.URL.createObjectURL = vi.fn();
    global.URL.revokeObjectURL = vi.fn();
    HTMLAnchorElement.prototype.click = vi.fn(); // Mock anchor DOM element
  });

  it('should match the snapshot', () => {
    const { container } = render(<PdfForm {...args} />);

    expect(container).toMatchSnapshot();

    expect(screen.getByRole('button', { name: args.button })).toBeInTheDocument();
  });

  it('should execute the form handlers when its submitted', async () => {
    render(<PdfForm {...args} />);

    fireEvent.click(screen.getByRole('button', { name: args.button }));

    await waitFor(() => expect(mockedCallback).toHaveBeenCalledOnce());

    expect(fetch).toHaveBeenCalledWith('/api/v1/pdf', {
      method: 'POST',
      next: { revalidate: 3600 },
    });
    expect(mockedCallback).toHaveBeenCalled();
  });

  it('should disable the button when loading', async () => {
    render(<PdfForm {...args} />);

    const button = screen.getByRole('button', { name: args.button });
    fireEvent.click(button);

    expect(button).toBeDisabled();
    expect(button).toHaveTextContent(args.buttonLoading);

    await waitFor(() => expect(mockedCallback).toHaveBeenCalledOnce());

    expect(button).not.toBeDisabled();
    expect(button).toHaveTextContent(args.button);
  });

  it('should handle fetch errors gracefully', async () => {
    global.fetch = vi.fn().mockImplementation(() =>
      Promise.resolve({
        status: 500,
        statusText: 'Internal Server Error',
      }),
    ) as Mock;

    const consoleSpy = vi.spyOn(console, 'error').mockImplementation(() => vi.fn());
    render(<PdfForm {...args} />);

    fireEvent.click(screen.getByRole('button', { name: args.button }));

    await waitFor(() =>
      expect(consoleSpy).toHaveBeenCalledWith('Request failed with status code 500'),
    );

    expect(mockedCallback).not.toHaveBeenCalled();

    consoleSpy.mockRestore();
  });
});
