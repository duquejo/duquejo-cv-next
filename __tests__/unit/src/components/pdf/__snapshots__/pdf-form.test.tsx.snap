// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<PdfForm /> tests > should match the snapshot 1`] = `
<div>
  <form
    aria-describedby="form-description"
    aria-labelledby="form-title"
    class="flex flex-col"
  >
    <div
      class="shrink-0 bg-border h-[1px] w-full mt-2 mb-3"
      data-orientation="horizontal"
      role="none"
    />
    <button
      class="inline-flex items-center justify-center gap-2 whitespace-nowrap font-medium transition-colors focus-visible:outline-hidden focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 bg-primary text-primary-foreground shadow-sm hover:bg-primary/90 h-8 rounded-md px-3 text-xs cursor-pointer disabled:cursor-not-allowed"
      type="submit"
    >
      Mocked button
    </button>
  </form>
</div>
`;
